From 9ffd7d3ba55a9dc63093f16a02b8d5634ffd2a39 Mon Sep 17 00:00:00 2001
From: Mahesh Bodapati <mbodapat@xilinx.com>
Date: Wed, 11 Aug 2021 15:31:47 +0530
Subject: [PATCH] [Patch,MicroBlaze] : Added zero_extendqidi2 and
 zero_extendhidi2 patterns to resolve TSR-974519. in GCC 10.2,commit
 0a237a94c206b53eb: Do not propagate results from inner REGS to paradoxical
 SUBREGs and this commit is causing the MB 64 compiler to generate extra
 instructions.

---
 gcc/config/microblaze/microblaze.md | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/gcc/config/microblaze/microblaze.md b/gcc/config/microblaze/microblaze.md
index 71ac46dfb6c..b34e963e176 100644
--- a/gcc/config/microblaze/microblaze.md
+++ b/gcc/config/microblaze/microblaze.md
@@ -1191,6 +1191,30 @@
   (set_attr "mode"	"SI,SI,SI")
   (set_attr "length"	"4,4,8")])
 
+(define_insn "zero_extendhidi2"
+  [(set (match_operand:DI 0 "register_operand" "=d,d,d")
+        (zero_extend:DI (match_operand:HI 1 "nonimmediate_operand" "d,R,m")))]
+  "TARGET_MB_64"
+  "@
+  andli\t%0,%1,0xffff
+  lhu%i1\t%0,%1
+  lhu%i1\t%0,%1"
+  [(set_attr "type"     "no_delay_arith,load,no_delay_load")
+  (set_attr "mode"      "DI,DI,DI")
+  (set_attr "length"    "8,4,8")])
+
+(define_insn "zero_extendqidi2"
+  [(set (match_operand:DI 0 "register_operand" "=d,d,d")
+        (zero_extend:DI (match_operand:QI 1 "nonimmediate_operand" "d,R,m")))]
+  "TARGET_MB_64"
+  "@
+  andli\t%0,%1,0x00ff
+  lbu%i1\t%0,%1
+  lbu%i1\t%0,%1"
+  [(set_attr "type"     "arith,load,no_delay_load")
+  (set_attr "mode"      "DI,DI,DI")
+  (set_attr "length"    "4,4,8")])
+
 ;;----------------------------------------------------------------
 ;; Sign extension
 ;;----------------------------------------------------------------
-- 
2.17.1

